File: backend/mr-db-sql/src/main/java/io/featurehub/db/services/UserStateSqlApi.java
package io.featurehub.db.services;

import cd.connect.app.config.ConfigKey;
import io.ebean.Database;
import io.featurehub.db.api.UserStateApi;
import io.featurehub.db.model.DbApplication;
import io.featurehub.db.model.DbPerson;
import io.featurehub.db.model.DbUserState;
import io.featurehub.db.model.UserState;
import io.featurehub.db.model.query.QDbEnvironment;
import io.featurehub.db.model.query.QDbUserState;
import io.featurehub.mr.model.HiddenEnvironments;
import io.featurehub.mr.model.Person;
import jakarta.inject.Inject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Collectors;

public class UserStateSqlApi implements UserStateApi {
  private static final Logger log = LoggerFactory.getLogger(UserStateSqlApi.class);
  private final Conversions conversions;
  private final Database database;
  @ConfigKey("limit.maxEnvironmentsPerApplication")
  protected Integer maximumEnvironmentsPerApplication = 1000;

  @Inject
  public UserStateSqlApi(Conversions conversions, Database database) {
    this.conversions = conversions;
    this.database = database;
  }

  private QDbUserState userStateFinder(Person person, UUID appUid, UserState userState) {
    Conversions.nonNullApplicationId(appUid);
    Conversions.nonNullPerson(person);

    return new QDbUserState()
      .person.id.eq(person.getId().getId())
      .application.id.eq(appUid)
      .userState.eq(userState);
  }

  @Override
  public HiddenEnvironments getHiddenEnvironments(Person person, UUID appUid) {
    Conversions.nonNullApplicationId(appUid);
    final DbUserState features = userStateFinder(person, appUid, UserState.HIDDEN_FEATURES).findOne();

    if (features != null) {
      return Conversions.readJsonValue(features.getData(), HiddenEnvironments.class);
    }

    return null;
  }

  @Override
  public void saveHiddenEnvironments(Person currentPerson, HiddenEnvironments environments, UUID appId) throws InvalidUserStateException {
    final DbApplication application = conversions.byApplication(appId);
    final DbPerson person = conversions.byPerson(currentPerson);

    if (application == null || person == null) {
      log.warn("Attempt made to save user state with invalid person or application.");
      return;
    }

    if (environments == null || environments.getEnvironmentIds() == null || environments.getEnvironmentIds().isEmpty()) {
      userStateFinder(currentPerson, application.getId(), UserState.HIDDEN_FEATURES).delete();
      return;
    }

    // too many environments?
    if (environments.getEnvironmentIds().size()  > maximumEnvironmentsPerApplication) {
      throw new InvalidUserStateException("Too many environments.");
    }

    // environment ids that aren't uuids?
    List<UUID> envIds =
      environments.getEnvironmentIds().stream().filter(Objects::nonNull).collect(Collectors.toList());

    if (envIds.size() != environments.getEnvironmentIds().size()) {
      throw  new InvalidUserStateException("Invalid UUIDs in environments list");
    }

    // environment ids that don't exist?
    if (new QDbEnvironment().id.in(envIds).findCount() != envIds.size()) {
      throw new InvalidUserStateException("Invalid number of environments in environments list");
    }

    DbUserState features =
        userStateFinder(currentPerson, application.getId(), UserState.HIDDEN_FEATURES).findOne();

    if (features == null) {
      features = new DbUserState.Builder()
          .application(application)
          .person(person)
          .userState(UserState.HIDDEN_FEATURES)
          .build();

      if (features.getApplication() == null || features.getPerson() == null ) {
        return;
      }
    }

    features.setData(Conversions.valueToJsonString(environments));

    if (features.getData() != null) {
      database.save(features);
    }
  }
}

File: backend/mr-db-sql/src/main/kotlin/io/featurehub/db/services/UserStateSqlApi.kt
package io.featurehub.db.services

import cd.connect.app.config.ConfigKey
import io.ebean.Database
import io.featurehub.db.api.UserStateApi
import io.featurehub.db.model.DbUserState
import io.featurehub.db.model.UserState
import io.featurehub.db.model.query.QDbEnvironment
import io.featurehub.db.model.query.QDbUserState
import io.featurehub.mr.model.HiddenEnvironments
import io.featurehub.mr.model.Person
import jakarta.inject.Inject
import org.slf4j.LoggerFactory
import java.lang.Boolean
import java.util.*
import java.util.stream.Collectors

class UserStateSqlApi @Inject constructor(private val conversions: Conversions) :
  UserStateApi {
  @ConfigKey("limit.maxEnvironmentsPerApplication")
  protected var maximumEnvironmentsPerApplication = 1000
  private fun userStateFinder(person: Person, appUid: UUID, userState: UserState): QDbUserState {
    Conversions.nonNullApplicationId(appUid)
    Conversions.nonNullPerson(person)
    return QDbUserState().person.id.eq(person.id!!.id).application.id.eq(appUid).userState.eq(userState)
  }

  override fun getHiddenEnvironments(person: Person, appUid: UUID): HiddenEnvironments? {
    Conversions.nonNullApplicationId(appUid)
    val features = userStateFinder(person, appUid, UserState.HIDDEN_FEATURES).findOne()
    return if (features != null) {
      Conversions.readJsonValue(
        features.data,
        HiddenEnvironments::class.java
      )!!
    } else null
  }

  @Throws(UserStateApi.InvalidUserStateException::class)
  override fun saveHiddenEnvironments(currentPerson: Person, environments: HiddenEnvironments, appId: UUID) {
    val application = conversions.byApplication(appId) ?: return
    val person = conversions.byPerson(currentPerson) ?: return

    if (environments.environmentIds.isEmpty() && (environments.noneSelected !== true)) {
      userStateFinder(currentPerson, application.id, UserState.HIDDEN_FEATURES).delete()
      return
    }

    // too many environments?
    if (environments.environmentIds.size > maximumEnvironmentsPerApplication) {
      throw UserStateApi.InvalidUserStateException("Too many environments.")
    }

    // environment ids that aren't uuids?
    val envIds = environments.environmentIds
      .stream().filter { obj: UUID? -> Objects.nonNull(obj) }.collect(Collectors.toList())
    if (envIds.size != environments.environmentIds.size) {
      throw UserStateApi.InvalidUserStateException("Invalid UUIDs in environments list")
    }

    // environment ids that don't exist?
    if (QDbEnvironment().id.`in`(envIds).findCount() != envIds.size) {
      throw UserStateApi.InvalidUserStateException("Invalid number of environments in environments list")
    }
    var features = userStateFinder(currentPerson, application.id, UserState.HIDDEN_FEATURES).findOne()
    if (features == null) {
      features = DbUserState.Builder()
        .application(application)
        .person(person)
        .userState(UserState.HIDDEN_FEATURES)
        .build()
    }

    features!!.data = Conversions.valueToJsonString(environments)

    if (features.data != null) {
      features.save()
    }
  }

  companion object {
    private val log = LoggerFactory.getLogger(UserStateSqlApi::class.java)
  }
}
